### PR TITLE
RAS monitor and clean up + DIR impl

### PR DESCRIPTION
**Pull request by Jimbo4794**
Signed-off-by: James Davies <jdavies47@hotmail.co.uk>

### CODE REVIEW

File: galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        this.ras = framework.getResultArchiveStore();
+    }
+
+    @Override
+    public void run() {
+        logger.info("Starting search for old Runs");
+
+        try {
+            long age = 0;
+            String units = cps.getProperty("ras", "cleanup.expire.unit");
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java:45]: sorry,  way overkill,   there is no way someone is going to request 30 seconds, or hours.   Only need to support days for this process.

File: galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunResourceManagement.java [hunk 2]
```diff
@@ -73,6 +73,13 @@ public void start() {
         } catch (FrameworkException e) {
             logger.error("Unable to initialise Finished Run monitor", e);
         }
+        try {
+            this.resourceManagement.getScheduledExecutorService().scheduleWithFixedDelay(
+                    new RunRASMonitor(this.framework, this.resourceManagement, this.dss, cps, this),
+                    this.framework.getRandom().nextInt(20), 20, TimeUnit.SECONDS);
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunResourceManagement.java:79]: We dont need this running every 20 seconds,   once an hour maybe

Comment (Michael-Baylis) [reply galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunResourceManagement.java:79]: and give it a long initial wait

File: galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                break;
+                case "MINUTES": age = (Duration.ofMinutes(amount).getSeconds());
+                break;
+                case "HOURS": age  = (Duration.ofHours(amount).getSeconds());
+                break;
+                case "DAYS": age  = (Duration.ofDays(amount).getSeconds());
+                break;
+            default: age  = (Duration.ofDays(21).getSeconds());
+            }
+            long now = Instant.now().getEpochSecond();
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java:58]: as days is the minimum interval,   you can use the instant plus methods.

**Review by Michael-Baylis — CHANGES_REQUESTED** [review]

File: galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunResourceManagement.java (hunk 2 above)
Comment (Jimbo4794) [reply galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunResourceManagement.java:79]: Ah that was left in from my testing

File: galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        this.cps = cps;
+        this.ras = framework.getResultArchiveStore();
+    }
+
+    @Override
+    public void run() {
+        logger.info("Starting search for old Runs");
+
+        try {
+            long amount = Long.parseLong(cps.getProperty("ras", "cleanup.days"));
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java:45]: What happens if cleanup.days is not provided,  should have a default

**Review by Michael-Baylis — CHANGES_REQUESTED** [review]

File: galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                // Set default value
+                cleanupDays = "21";
+            }
+            
+
+            long amount = Long.parseLong(cleanupDays);
+            long cleanupAge = Duration.ofDays(amount).getSeconds();
+            long now = Instant.now().getEpochSecond();
+            
+            Instant requestUpTo = Instant.ofEpochSecond(Math.subtractExact(now, age));
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java:57]: why use epochseconds.   use Instand.plus(ChronoUnit.Days, xxxx);

File: galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        try {
+            long age = 0;
+            String cleanupDays = cps.getProperty("ras.cleanup", "days");
+            if (cleanupDays == null) {
+                // Set default value
+                cleanupDays = "21";
+            }
+            
+
+            long amount = Long.parseLong(cleanupDays);
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java:53]: need to throw a nice exception if cleanupDays can't be parsed

File: galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            long cleanupAge = Duration.ofDays(amount).getSeconds();
+            long now = Instant.now().getEpochSecond();
+            
+            Instant requestUpTo = Instant.ofEpochSecond(Math.subtractExact(now, age));
+
+            IResultArchiveStore ras = framework.getResultArchiveStore();
+            for(IResultArchiveStoreDirectoryService service :  ras.getDirectoryServices()) {
+                for (IRunResult run : service.getRuns(new RasSearchCriteriaQueuedTo(requestUpTo))) {
+                    logger.info("Discarding old run: " + run.getTestStructure().getRunName());
+                    run.discard();
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.resource.management/src/main/java/dev/galasa/framework/resource/management/internal/RunRASMonitor.java:63]: surround this with a try block.   so if a run can't be discard,  it doesn't stop other runs from being discarded.

**Review by Michael-Baylis — CHANGES_REQUESTED** [review]

**Review by Michael-Baylis — CHANGES_REQUESTED** [review]
you didnt action the 
"need to throw a nice exception if cleanupDays can't be parsed"

**Comment by Jimbo4794** [comment]
There is a seperate exception thrown for the NumberFormatException on line 67

**Pull request closed without merging**